Golang-style salt lengths to verify RSA PSS sigs

trishankatdatadog opened the pull request:
Signed-off-by: Trishank Karthik Kuppusamy <trishank.kuppusamy@datadoghq.com>

**Fixes issue #**:

N/A

**Description of the changes being introduced by the pull request**:

Golang-style salt lengths to allow us to verify RSA PSS signatures produced in Go-based software such as [Hashicorp Vault](https://github.com/hashicorp/vault/blob/f6547fa8e820b6ebbfa15018477a138b38707d91/sdk/helper/keysutil/policy.go#L1363).

**Please verify and check that the pull request fulfils the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

shibumi commented on securesystemslib/rsa_keys.py:
```diff
@@ -473,9 +481,18 @@ def verify_rsa_signature(signature, signature_scheme, public_key, data):
     # hashing algorithm.
     try:
       if signature_scheme.startswith('rsassa-pss'):
+        # https://github.com/golang/go/blob/11f92e9dae96939c2d784ae963fa7763c300660b/src/crypto/rsa/pss.go#L269-L275
```
Can you add a sentence or two, why you add this link to Go code here?
Also what is `padding.PSS.MAX_LENGTH`? For me, it sounds like that the padding determines the pss salt length.
Is this correct?

shibumi commented on securesystemslib/rsa_keys.py:
```diff
@@ -486,7 +503,7 @@ def verify_rsa_signature(signature, signature_scheme, public_key, data):
       # This is a defensive check check..
       else:  # pragma: no cover
         raise securesystemslib.exceptions.UnsupportedAlgorithmError('Unsupported'
-            ' signature scheme is specified: ' + repr(scheme))
```
Where is this 'scheme' coming from? I don't see that you removed or renamed scheme in another line. Or was this done in a previous commit, unrelated to this PR?

trishankatdatadog replied:
Exactly, it was a [bug](https://github.com/secure-systems-lab/securesystemslib/pull/262/files#r464378530).

shibumi commented:
> [![Coverage Status](https://camo.githubusercontent.com/8e0b9bf6c2907e392b19be69fba6f44b8025d53f/68747470733a2f2f636f766572616c6c732e696f2f6275696c64732f33323232393738322f6261646765)](https://coveralls.io/builds/32229782)
> 
> Coverage decreased (-98.9%) to 0.0% when pulling **[d148b64](https://github.com/secure-systems-lab/securesystemslib/commit/d148b6424111efdac26bcc0c1213ee382b857ea7) on trishankatdatadog:trishankatdatadog/options-for-rsa-pss-salt-length** into **[6f88f63](https://github.com/secure-systems-lab/securesystemslib/commit/6f88f63d1b4ed108c2abbc966b265d3dce42c8ea) on secure-systems-lab:master**.

Lol.. this sounds odd.

joshuagl commented on securesystemslib/rsa_keys.py:
```diff
@@ -382,7 +382,20 @@ def create_rsa_signature(private_key, data, scheme='rsassa-pss-sha256'):
+
+
+
+def verify_rsa_signature(signature, signature_scheme, public_key, data, salt_length=SaltLength.PSSSaltLengthEqualsHash):
```
Please document the new `salt_length` argument.

joshuagl commented on securesystemslib/rsa_keys.py:
```diff
@@ -455,11 +468,6 @@ def verify_rsa_signature(signature, signature_scheme, public_key, data):
   # What about 'data'?
   securesystemslib.formats.DATA_SCHEMA.check_match(data)
 
-  # Verify whether the private key of 'public_key' produced 'signature'.
-  # Before returning the 'valid_signature' Boolean result, ensure 'RSASSA-PSS'
-  # was used as the signature scheme.
-  valid_signature = False
-
```
💯

joshuagl commented on securesystemslib/rsa_keys.py:
```diff
@@ -486,7 +503,7 @@ def verify_rsa_signature(signature, signature_scheme, public_key, data):
       # This is a defensive check check..
       else:  # pragma: no cover
         raise securesystemslib.exceptions.UnsupportedAlgorithmError('Unsupported'
-            ' signature scheme is specified: ' + repr(scheme))
```
Thanks for catching this.

trishankatdatadog commented:
@shibumi @joshuagl Addressed your comments by adding tests and documentation.

I'm not happy with fixing the RSA-PSS salt length in the code, instead of as part of signed metadata on how to use the public key. We need to design a more sensible key distribution format going forward instead of using a single `scheme` string to encode hashing+signing+marshaling algorithms.

joshuagl commented on securesystemslib/rsa_keys.py:
```diff
@@ -246,7 +246,28 @@ def generate_rsa_public_and_private(bits=_DEFAULT_RSA_KEY_BITS):
+
+
+
+def create_rsa_signature(private_key, data, scheme='rsassa-pss-sha256', salt_length=SaltLength.PSSSaltLengthEqualsHash):
```
Coding style requires we wrap at 79 characters and double-indent the continued line.

shibumi replied:
How did this even pass the test? I thought our travis checks for linewrapping? I made another PR a few days ago and there travis complained about lines longer than 79 characters.. weird

lukpueh replied:
No linting on seecuresystemslab yet (#243). Maybe your PR was on in-toto?

shibumi replied:
@lukpueh ah yes.. sorry for the noise :) my mistake

joshuagl commented on securesystemslib/rsa_keys.py:
```diff
@@ -382,7 +419,7 @@ def create_rsa_signature(private_key, data, scheme='rsassa-pss-sha256'):
 
 
-def verify_rsa_signature(signature, signature_scheme, public_key, data):
+def verify_rsa_signature(signature, signature_scheme, public_key, data, salt_length=SaltLength.PSSSaltLengthEqualsHash):
```
Please wrap this line too.

joshuagl commented on securesystemslib/rsa_keys.py:
```diff
@@ -334,12 +361,22 @@ def create_rsa_signature(private_key, data, scheme='rsassa-pss-sha256'):
         'pyca_crypto')
 
     if scheme.startswith('rsassa-pss'):
+      if salt_length == SaltLength.PSSSaltLengthAuto:
+        salt_length = padding.PSS.MAX_LENGTH
+      elif salt_length == SaltLength.PSSSaltLengthEqualsHash:
+        salt_length = digest_obj.algorithm.digest_size
+      else:
+        raise securesystemslib.exceptions.UnsupportedAlgorithmError('Unsupported'
+          ' salt length is specified: ' + repr(salt_length))
```
Can we have different variable names for the passed salt_length (of type SaltLength) and the salt_length we are passing to `sign()` (presumably an int?)? Reusing variable names can be tricky to read and lead to unexpected errors later.

joshuagl commented on securesystemslib/rsa_keys.py:
```diff
@@ -473,9 +511,18 @@ def verify_rsa_signature(signature, signature_scheme, public_key, data):
     # hashing algorithm.
     try:
       if signature_scheme.startswith('rsassa-pss'):
+        if salt_length == SaltLength.PSSSaltLengthAuto:
+          salt_length = padding.PSS.MAX_LENGTH
+        elif salt_length == SaltLength.PSSSaltLengthEqualsHash:
+          salt_length = digest_obj.algorithm.digest_size
+        else:
+          raise securesystemslib.exceptions.UnsupportedAlgorithmError('Unsupported'
+            ' salt length is specified: ' + repr(salt_length))
```
Perhaps we could factor this logic out into a helper, as we use it here in verify and earlier in create.

joshuagl commented on tests/test_rsa_keys.py:
```diff
@@ -66,46 +66,54 @@ def test_create_rsa_signature(self):
-          data)
+      if rsa_scheme.startswith('rsassa-pss'):
+        self.assertRaises(securesystemslib.exceptions.UnsupportedAlgorithmError,
+          securesystemslib.rsa_keys.create_rsa_signature, private_rsa, data, rsa_scheme, salt_length=123)
```
Please wrap this line

joshuagl commented on tests/test_rsa_keys.py:
```diff
@@ -66,46 +66,54 @@ def test_create_rsa_signature(self):
+        securesystemslib.rsa_keys.SaltLength.PSSSaltLengthEqualsHash,
+      ):
+        signature, scheme = \
+          securesystemslib.rsa_keys.create_rsa_signature(private_rsa, data, rsa_scheme, salt_length=salt_length)
```
Could you wrap this line too?

joshuagl commented on tests/test_rsa_keys.py:
```diff
@@ -114,52 +122,62 @@ def test_verify_rsa_signature(self):
+        securesystemslib.rsa_keys.SaltLength.PSSSaltLengthEqualsHash,
+      ):
+        signature, scheme = \
+          securesystemslib.rsa_keys.create_rsa_signature(private_rsa, data, rsa_scheme, salt_length=salt_length)
```
Another line in need of wrapping, please

joshuagl commented on securesystemslib/rsa_keys.py:
```diff
@@ -474,19 +528,22 @@ def verify_rsa_signature(signature, signature_scheme, public_key, data):
-      # The RSA_SCHEME_SCHEMA.check_match() above should have validated 'scheme'.
-      # This is a defensive check check..
+      # The RSA_SCHEME_SCHEMA.check_match() above should have validated
+      # 'scheme'. This is a defensive check check..
```
```suggestion
      # 'scheme'. This is a defensive check.
```

joshuagl commented on tox.ini:
```diff
@@ -42,3 +42,13 @@ commands =
+    # https://github.com/secure-systems-lab/sample-documents/blob/master/pylintrc
+    # FIXME: Run pylint on everything except _vendor. Also, run bandit.
+    pylint securesystemslib/rsa_keys.py --disable C0103
+    pylint tests/test_rsa_keys.py --disable C0103,C0115,C0116,W0212,W0603
```
Nit: no newline at the end of the file

trishankatdatadog commented on tests/test_rsa_keys.py:
```diff
@@ -66,100 +68,110 @@ def test_create_rsa_signature(self):
-          FORMAT_ERROR_MSG)
-      self.assertEqual(rsa_scheme, scheme)
+      for salt_length_type in (
+        securesystemslib.rsa_keys.HashSaltLengthType,
+        securesystemslib.rsa_keys.MaxSaltLengthType,
```
GitHub says there is no difference here...

trishankatdatadog commented on securesystemslib/rsa_keys.py:
```diff
@@ -415,6 +471,9 @@ def verify_rsa_signature(signature, signature_scheme, public_key, data):
       'signature'.  'data' (a string) is needed here to verify 'signature'.
 
+    salt_length_type:
+      The length of salt used in RSA-PSS.
```
I want to reduce the scope for error, so using types that will choose one of two sane possibilities seems more reasonable to me.

joshuagl replied:
Agreed, I really like the types. I was meaning to imply that the docstring might be read as suggesting an int length. 
```suggestion
      The strategy for determining the length of the salt used in RSA-PSS, one of
      HashSaltLengthType or MaxSaltLengthType.
```

trishankatdatadog commented:
Done. Let me know if there are major issues. Any major objections @lukpueh?

lukpueh reviewed (commented):
Many thanks for the enhancements, @trishankatdatadog! Here are some high-level observations:
- Tests need to pass
- Could you please move the linter configuration and hence resulting style/format fixes to a separate PR?
- Could you please squash your commits as appropriate and add a bit more info about the changes to the commit messages?
- There should be no merge commits in PRs.

lukpueh commented:
Thanks for the updates! More details in the commit message would still be nice.

trishankatdatadog commented:
> Thanks for the updates! More details in the commit message would still be nice.

Finally done! Please take a last look...

trishankatdatadog commented:
Any news, please? @lukpueh

lukpueh commented on securesystemslib/rsa_keys.py:
```diff
@@ -245,7 +247,54 @@ def generate_rsa_public_and_private(bits=_DEFAULT_RSA_KEY_BITS):
+# its crypto package.
+# https://github.com/golang/go/blob/11f92e9dae96939c2d784ae963fa7763c300660b/src/crypto/rsa/pss.go#L225-L232
+# FIXME: really, we should encode the salt length as part of the metadata on
+# how to use the RSA-PSS public key.
```
👍  Agreed. Maybe configured via the "scheme" field? See #308.

lukpueh commented on securesystemslib/rsa_keys.py:
```diff
@@ -245,7 +247,54 @@ def generate_rsa_public_and_private(bits=_DEFAULT_RSA_KEY_BITS):
+# https://github.com/golang/go/blob/11f92e9dae96939c2d784ae963fa7763c300660b/src/crypto/rsa/pss.go#L225-L232
+# FIXME: really, we should encode the salt length as part of the metadata on
+# how to use the RSA-PSS public key.
+class SaltLengthType:
+  """A class to represent common salt lengths for RSA-PSS."""
+
+  @classmethod
+  def get_salt_length(self, algorithm):
+    """Get the salt length as integer."""
+    raise NotImplementedError
+
+
+
+
+
+# NOTE: This is what used to be the standard behaviour.
+class HashSaltLengthType(SaltLengthType):
+  """Salt length to equal the length of the hash used in the signature."""
+
+  @classmethod
+  def get_salt_length(cls, algorithm):
+    """Get the salt length as integer."""
+    return algorithm.digest_size
+
+
+
+
+
+class MaxSaltLengthType(SaltLengthType):
+  """Salt length in a PSS signature to be as large as possible when
+  signing, and to be auto-detected when verifying."""
+
+  @classmethod
+  def get_salt_length(cls, algorithm):
+    """Get the salt length as integer."""
+    # NOTE: We disregard algorithm here.
+    return padding.PSS.MAX_LENGTH
```
Smart architecture. :) Do we expect other `SaltLengthType` implementations that warrant an interface? Wouldn't a simple `if/elif` switch akin to the [go code you referenced](https://github.com/golang/go/blob/11f92e9dae96939c2d784ae963fa7763c300660b/src/crypto/rsa/pss.go#L269-L275) be just as good and a little less "engineered"?

lukpueh commented on securesystemslib/rsa_keys.py:
```diff
@@ -245,7 +247,54 @@ def generate_rsa_public_and_private(bits=_DEFAULT_RSA_KEY_BITS):
+
+
+def create_rsa_signature(private_key, data, scheme='rsassa-pss-sha256',
+  salt_length_type=HashSaltLengthType):
```
I'm a bit worried how this extra parameter will fit in with the regular securesystemslib usage patterns, where algorithmic specific sign functions are only called indirectly via `securesystemslib.keys.create_signature`, dispatching based on the signing scheme field of the private key. So updating encoding the salt length type, as you suggested above, is probably the way to go. But we can do this in a follow-up PR.

lukpueh commented on securesystemslib/rsa_keys.py:
```diff
@@ -278,6 +327,10 @@ def create_rsa_signature(private_key, data, scheme='rsassa-pss-sha256'):
       The signature scheme used to generate the signature.
 
+    salt_length_type:
+      The strategy for determining the length of the salt used in RSA-PSS, one of
```
It's a bit ironic that there are plenty of unrelated re-wraps in your diff to match the prescribed line length, but a line that you added exceeds it.

I wish it was the other way around. Could you please remove non-related style changes from this PR as requested in [an earlier review](https://github.com/secure-systems-lab/securesystemslib/pull/262#pullrequestreview-471518591).

lukpueh commented on securesystemslib/rsa_keys.py:
```diff
@@ -381,7 +437,8 @@ def create_rsa_signature(private_key, data, scheme='rsassa-pss-sha256'):
 
-def verify_rsa_signature(signature, signature_scheme, public_key, data):
+def verify_rsa_signature(signature, signature_scheme, public_key, data,
+  salt_length_type=HashSaltLengthType):
```
Same as above, we usually just call `securesystemslib.keys.verify_signature` and use the key to dispatch to the algorithm specific function.

lukpueh reviewed (changes requested):
Apologies for the late review. I think this looks quite good. I would probably implement it a bit differently, but I guess that's a matter of taste. Please remove unrelated changes and I can give it another pass.

trishankatdatadog commented:
Thanks, Lukas, but I'm afraid I don't have the bandwidth to work on this PR anymore, especially since there are no plans to upstream [my other PR](https://github.com/theupdateframework/tuf/pull/1060) which depends on this one. Should we just close this for now and revisit in the future?

trishankatdatadog commented:
Note to @woodruffw though: you will need something like this PR to finish your Vault-TUF (or is it Vault-securesystemslib) integration. Closing for now.
